### PR TITLE
Gsdx half pixel offset new

### PR DESCRIPTION
**Pull request by gregory38**
Tentative to implement the HPO hack differently. I still see some rainbow on gow2 that I need to debug further. But so far results look interesting. Feel free to give it a try on ghosting effect due to upscaling.

```
UserHacks_HalfPixelOffset_New = 1
```

### CODE REVIEW

**Comment by gregory38** [comment]
Some idea to investigate
* use the likely offset hint ?
* don't check texture position ?

**Comment by Blackbird88** [comment]
Old
image
New
image

**Comment by gregory38** [comment]
Well goal is to fix some border/ghosting issues. I don't see any on MGS3 ;)

**Comment by Blackbird88** [comment]
Well at least you know there are no regressions :)
This is how it looks without any hack
image

**Comment by ramapcsx2** [comment]
This might be a good topic to discuss the Wild Arms Offset ;p
That older hack is tailored for maybe 3 games in the Wild Arms series and on everything else, the result is pretty random.
The other hacks that handle alignment should be able to handle Wild Arms games as well.
So I tested them and it turns out that the Round Sprite hack (either half or full) almost works for this.
It fixes the horizontal glitches but unfortunately creates blur on the verticals.
Is the Round Sprite hack only working differently on the x and y axis?

**Comment by FlatOutPS2** [comment]
Great. This also fixes an issue of gaps between sprites/lines accross the screen in several 2D fighters!

**Comment by gregory38** [comment]
@ramapcsx2 what do you mean about the de-interlace? The hack mustn't have any major impact on it.

@FlatOutPS2 hum, not expected but possible, could you give me a GS dump ?

**Comment by ramapcsx2** [comment]
@gregory38 Somehow it does though. Just enabling the Round Sprite hack blurs vertically in this game.

**Comment by gregory38** [comment]
Hum, I guess the game blit the target to the crtc buffer. Rounding will somehow create a slight distortion. It is annoying.

**Comment by ramapcsx2** [comment]
Okay, then it's probably the same thing as in Everblue and should be handled in a separate issue.
For the record: This is how Everblue always looked / still looks in HW, SW, native, whatever:
http://i.imgur.com/gRxV08M.jpg
Fiddling with offsets in the hacks section affects it.

**Comment by ssakash** [comment]
@ramapcsx2 

Sorry for being off topic but could you provide a GS dump of the issue ? I'd love to analyze it ;)

**Comment by gregory38** [comment]
Arg you burn my eye :p It is awful.

**Comment by ramapcsx2** [comment]
https://www.sendspace.com/file/4prj6z

**Comment by FlatOutPS2** [comment]
[Samurai Shodown.zip](https://github.com/PCSX2/pcsx2/files/326187/Samurai.Shodown.zip)

2D Fighter lines issue example GS Dump.

**Comment by gregory38** [comment]
@FlatOutPS2 Did you use an extra setting, the new hack doesn't improve the situation here ? (tested on 4x)

**Comment by FlatOutPS2** [comment]
No, no extra settings. I just discovered it was merely a coincidence though. The games switches between interlaced and progressive a lot(might the at the source of the issue), and it sometimes creates or removes the issue on pause/reboots. And when I added and removed the new HPO several times it just happened to switch as well.

**Comment by refractionpcsx2** [comment]
> This is how Everblue always looked / still looks in HW, SW, native, whatever

Everyblue has always shaken violently though :P

**Comment by tabnk** [comment]
May fix Tekken 5 too :)

**Comment by tabnk** [comment]
Haunting Ground

 UserHacks_HalfPixelOffset_New = 0
gsdx_20160623165756


UserHacks_HalfPixelOffset_New = 1
gsdx_20160623165449

**Comment by refractionpcsx2** [comment]
> Haunting Ground
> 
> UserHacks_HalfPixelOffset_New = 0

What about the old half pixel offset?

**Comment by tabnk** [comment]
Only hack enabled was UserHacks_HalfPixelOffset_New = 1 for Haunting Ground. 

Didn't test old half pixel offset.

**Comment by refractionpcsx2** [comment]
> Didn't test old half pixel offset.

it would be good if you could to show if there's any improvement or not.

**Comment by tabnk** [comment]
Haunting Ground
With old half pixel offset enabled 
gsdx_20160623184023

**Comment by FlatOutPS2** [comment]
I'd say the old one looks ever so slighty better than the new one.

**Comment by refractionpcsx2** [comment]
the old one looks much better. If you open them both in new tabs (the new and old) the new one still has ghosting.

**Comment by ramapcsx2** [comment]
UserHacks_HalfPixelOffset_New isn't quite as effective as "old half pixel offset" but there's no problem on the left edge of the screen. Offset hacks often create visible vertical traces on the left side.
If UserHacks_HalfPixelOffset_New fixes those, it's definitely better :)

**Comment by refractionpcsx2** [comment]
> UserHacks_HalfPixelOffset_New isn't quite as effective as "old half pixel offset" but there's no problem on the left edge of the screen. Offset hacks often create visible vertical traces on the left side.
> If UserHacks_HalfPixelOffset_New fixes those, it's definitely better :)

but has it fixed that problem at the expense of the rest of he picture? Id rather there was a glitch down the left hand side of the screen than all over the screen tbh :P

**Comment by gregory38** [comment]
Well there is a catch. The way, I implemented the hack doesn't remove the GS effect. Instead it applies it at the correct location. Do you see the ghosting in native resolution ?
In short old hack is
```
add a 0.5 offset to vertex position
```
The new hack
```
+		if (!linear && t_position == 8) {
+			half_offset = 8 - 8 / scale.x;
+		} else if (linear && t_position == 16) {
+			half_offset = 16 - 16 / scale.x;
+		}
```
In plain english
```
remove 0.5 offset to texture position
add back 0.5 / upscaling offset to texture position
```
So at 2x, you have a shift of 0.25 instead of 0.5

**Comment by refractionpcsx2** [comment]
So are you saying it is intended?

**Comment by ramapcsx2** [comment]
It's definitely more fine grained, which might help some games better than others.
There's probably no "one size fits all" hack to align these blending effects.

**Comment by gregory38** [comment]
@refractionpcsx2 yeah more or less. Well for me it is more mathematically correct, it doesn't mean that it is a nicer image. Honestly the hack can still be tuned, it was only a trial.

What happen if you remove the negative offset part ` - 8 / scale.x` and ` - 16 / scale.x` ?

**Comment by ramapcsx2** [comment]
Tales of the Abyss seems to require the old (whole texel moving) hack to fix the effects.

Without hacks:
http://i.imgur.com/JPUZjW0.jpg

With old hack enabled:
http://i.imgur.com/S858qcf.jpg

With new hack enabled:
http://i.imgur.com/Ap0dfuf.jpg

**Comment by gregory38** [comment]
Hum I need to check my gs dump, I didn't see those ghosts. I'm likely blind :)

**Comment by gregory38** [comment]
@ramapcsx2 yes give me a dump. Please. I have another angle which seems to work.

**Comment by ramapcsx2** [comment]
@gregory38 here you go: https://www.sendspace.com/file/0hcmsh
:)

**Comment by gregory38** [comment]
Could you post your gsdx.ini file too. I don't see those ghosting in the dump.

**Comment by ramapcsx2** [comment]
My settings are all default. You get the blur by upscaling (3x in my case).

**Comment by FlatOutPS2** [comment]
I don't see the ghosting on the dump either, no matter how high the upscaling.

**Comment by ramapcsx2** [comment]
Ah, sorry. This blur is crc hacked. You need to set the crc level to disabled.

**Comment by gregory38** [comment]
Oh the small detail :p

**Comment by gregory38** [comment]
I pushed a new version that does a -0.5f offset on texture for float coordinate. Tell me if it brokes the game.

**Comment by gregory38** [comment]
@tabnk 
Could you test UserHacks_HalfPixelOffset_New = 2 on haunting ground ?

**Comment by tabnk** [comment]
Haunting Ground
**UserHacks_HalfPixelOffset_New = 2**

gsdx_20160625084500

GS dump if you need it.
https://www.sendspace.com/file/tobka1

**Comment by woj1993** [comment]
@gregory38 This does not help ghosting issues in Okami and Captain Tsubasa. Do you want a gs dump from these games? Gs dump ale legal to post, right? Sorry if I make any mistakes, I'm not native English speaker and sorry if it is not the right place to write this.

**Comment by gregory38** [comment]
Yes, could you send me some dumps. I'm not a layer but gs dump ought to be legal (there aren't playable, only debug GSdx code).

**Comment by woj1993** [comment]
With screenshot of native resolution. Good luck.
https://www.dropbox.com/sh/akr3lurkp1hgx4p/AAAaP-IgBi3J0CXLddjrlsQ6a?dl=0

**Comment by FlatOutPS2** [comment]
The second Captain Tsubasa dump(ghost issue 2) is actually another dump from Okami.

**Comment by woj1993** [comment]
Sorry my error. Tsubasa ghosting issues 2 is in it own archive and "Ghosting issues New" have correctly named second file from okami without screenshot.

**Comment by FlatOutPS2** [comment]
@woj1993 Can you make a slightly longer Okami dump, like 1-2 seconds long(keep holding Control+Shift+F8 for 1-2 seconds)?

**Comment by woj1993** [comment]
@FlatOutPS2 Can you check tchem? The same link, file name "Okami ghosting issue long dumps".

**Comment by FlatOutPS2** [comment]
Thanks for the dumps, just the right length I needed. I already fixed the Captain Tsubasa blur from the dump in a custom build, but it doesn't seem to affect Okami. But they're completely different issues, so that was to be expected. :p

**Comment by gregory38** [comment]
I'm currently looking at Captain Tsubasa. The last draw calls blur the screen (border), you can see the effect it in native too. However the blurring is 1 pixel in native and 2-3 in 2x. I need to debug further.

**Comment by gregory38** [comment]
Hum before I forget, so the game apply a -0.5 offset on the vertex position, texture doesn't have an offset. I guess we need to add a 0.5 offset (or maybe less) if first pixel is in a negative position.

**Comment by FlatOutPS2** [comment]
captain

**Comment by gregory38** [comment]
Oh and what did you do ?

**Comment by FlatOutPS2** [comment]
I made a different new half pixel offset. :p

**Comment by refractionpcsx2** [comment]
> I made a different new half pixel offset. :p

code? :P

**Comment by FlatOutPS2** [comment]
> code? :P

Yes, in code :p

I had some VS issues yesterday, I'll have to rebuild it and test it on more games to see if the code is worth using or just lucky garbage.

**Comment by refractionpcsx2** [comment]
> Yes, in code :p

*facepalm* I mean do you care to share your code please so we can see what you did? :P You've said you've done it but given no clue to how you did it lol

**Comment by FlatOutPS2** [comment]
> *facepalm* I mean do you care to share your code please so we can see what you did? :P You've said you've done it but given no clue to how you did it lol

I have no clue how I did it. :P That's not entirely true, but it is based on some guesswork. I'd like to understand it better myself and test it before sharing some potentially useless code. :P And I'll have to check up on the code if it's still the same state I left it in after VS decided to f*** with it.

**Comment by refractionpcsx2** [comment]
Fair enough :P Why didn't you just say that? lol

**Comment by gregory38** [comment]
Normally if you set a negative half-offset on the vertex, you're good. But you might need my help to enable it only when the draw call needs it.

**Comment by gregory38** [comment]
@FlatOutPS2 could you show me your code ? I won't bite promise :stuck_out_tongue:

**Comment by woj1993** [comment]
@gregory38 Can I ask: "any progress"?

**Comment by gregory38** [comment]
no unfortunately. It isn't easy to debug, so I did different task (didn't have much free time neither).

**Comment by refractionpcsx2** [comment]
it would be helpful if @FlatOutPS2 posted his code, even in private :P

**Comment by gregory38** [comment]
So I don't know what @FlatOutPS2 did. But I added an extra offset for captain tsubasa when vertex position is negative.

**Comment by woj1993** [comment]
@gregory38 Thanks for fix to Capitan Tsubasa. I only ask what about Okami?
@FlatOutPS2 Can you post your code?

**Comment by refractionpcsx2** [comment]
well did it help okami?

**Comment by gregory38** [comment]
No, Okami does a ghosting effect on purpose. You can see it on native too. Effect is "bigger" with upscaling.
I saw a -1.5 vertex offset on some draw calls. Maybe removing the offset could help to reduce the issue. It would be an agressive hack.

**Comment by woj1993** [comment]
@gregory38 Yeah you are right. Sorry about my mistake. I was thinking that there is problem in ghosting because after upscaling women shadow was stretched out. One more thing you probably considered it but I think that proper way of upscaling will be if you calculate offset of "effect" and do +/-(offset) * upscaling value but I might be wrong. PS. What stop this branch from begin merged?

**Comment by gregory38** [comment]
Hum, I need to check the performance impact. It add a couple of extra instruction in the Vertex Shaders. And need to check again tales of abyss.

**Comment by woj1993** [comment]
@gregory38  Is this PR finished?
P.S. Can you look at this Issue report https://github.com/PCSX2/pcsx2/issues/1558 . Is it good?

**Comment by refractionpcsx2** [comment]
> P.S. Can you look at this Issue report #1558 . Is it good?

That looks to be unrelated to this PR as it affects software mode too.

**Comment by woj1993** [comment]
@refractionpcsx2 You are right. I'm only wanted to know if it is proper Bug report and what is with this branch. About this error did I should mention it in [Meta] issues with hardware or software?

**Comment by refractionpcsx2** [comment]
> @refractionpcsx2 You are right. I'm only wanted to know if it is proper Bug report and what is with this branch. About this error did I should mention it in [Meta] issues with hardware or software?

If you want to get gregory's attention, ask on the report you made or send him a PM on the forum, let's try and keep the discussion relevant. If you start tagging other issues/PR's it can get confusing to what is related to what.

**Comment by gregory38** [comment]
Gregory is busy to prepare his vacation. He'll leave soon without any internet connection.

And don't worry he follows all the project and his aware of the tons of work. No need to ping it, mailbox is full already ;)

**Comment by FlatOutPS2** [comment]
>Gregory is busy to prepare his vacation. He'll leave soon without any internet connection.

Will you be filling in for Gregory? :)

**Comment by gregory38** [comment]
Not gonna happen. My master will cut me down electricity to reduce the bills.
